File: src/maskgw/admin/ui/protocol.py
"""Protocolo declarativo fechado, sem HTTP, runtime ou efeitos de startup."""

from __future__ import annotations

import json
import re
from contextlib import suppress
from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field, ValidationError

MAX_DEPTH = 16
MAX_MODELS = 128
MAX_CONTROLS = 512
MAX_PRESENTATION_BYTES = 262144
SAFE_INTEGER = 9007199254740991
FORBIDDEN = frozenset({"__proto__", "constructor", "prototype"})
Scalar = str | bool | int | None
Segment = str | int


class InvalidPresentationError(ValueError):
    """Erro publico fixo; nao contem entrada rejeitada."""


class Closed(BaseModel):
    model_config = ConfigDict(extra="forbid", strict=True, frozen=True)


class Text(Closed):
    type: Literal["string"]
    min: int = Field(default=0, ge=0, le=SAFE_INTEGER)
    max: int = Field(default=SAFE_INTEGER, ge=0, le=SAFE_INTEGER)
    prefix: str = ""
    alphabet: str = ""


class Integer(Closed):
    type: Literal["integer"]
    min: int = Field(default=0, ge=0, le=SAFE_INTEGER)
    max: int = Field(default=SAFE_INTEGER, ge=0, le=SAFE_INTEGER)


class Boolean(Closed):
    type: Literal["boolean"]


class Choice(Closed):
    type: Literal["enum"]
    choices: list[str | int | bool] = Field(min_length=1)


class FieldLink(Closed):
    name: str
    ref: str
    required: bool
    default: Scalar = None


class Object(Closed):
    type: Literal["object"]
    fields: list[FieldLink]


class Sequence(Closed):
    type: Literal["list", "nullable"]
    item: str


class Variant(Closed):
    value: str | bool | int
    ref: str


class Union(Closed):
    type: Literal["union"]
    tag: str
    variants: list[Variant] = Field(min_length=1)


Shape = Annotated[
    Text | Integer | Boolean | Choice | Object | Sequence | Union, Field(discriminator="type")
]


class Definition(Closed):
    id: str
    shape: Shape


class Call(Closed):
    id: str
    method: Literal["GET", "POST", "PUT", "DELETE"]
    path: str
    input: str | None
    output: str
    error: str
    operation: Literal["read", "create", "replace", "delete", "move", "check", "confirm", "append"]
    identity: str | None


class Condition(Closed):
    type: Literal["present", "equal", "choice", "boolean"]
    path: list[Segment]
    value: Scalar = None


class Projection(Closed):
    type: Literal["copy", "object", "list", "omit", "insert", "replace", "remove", "permute"]
    source: Literal["base", "draft"]
    paths: list[list[Segment]] = Field(max_length=MAX_CONTROLS)
    target: list[Segment]


class Control(Closed):
    id: str
    type: Literal["text", "integer", "checkbox", "select", "table", "list", "read", "confirm"]
    path: list[Segment]
    model: str
    label: str
    default: Scalar = None
    condition: Condition | None = None
    projections: list[Projection] = Field(default_factory=list, max_length=MAX_CONTROLS)


class View(Closed):
    id: str
    label: str
    call: str
    controls: list[Control]
    actions: list[str]


class Editor(Closed):
    id: str
    name: str
    model: str
    controls: list[Control]
    help: str


class Binding(Closed):
    id: str
    role: Literal["version", "identity", "order", "consent", "result"]
    model: str
    path: list[Segment]


class Message(Closed):
    id: str
    name: str
    text: str
    state: Literal[
        "loading",
        "authentication",
        "draft",
        "pending",
        "success",
        "conflict",
        "busy",
        "incompatible",
        "unknown",
        "uncertain",
    ]


class Presentation(Closed):
    format: Literal[1]
    models: list[Definition] = Field(min_length=1, max_length=MAX_MODELS)
    calls: list[Call] = Field(min_length=19, max_length=19)
    views: list[View] = Field(min_length=6, max_length=6)
    editors: list[Editor] = Field(min_length=8, max_length=8)
    bindings: list[Binding]
    messages: list[Message]


def _pairs(pairs: list[tuple[str, object]]) -> dict[str, object]:
    result: dict[str, object] = {}
    for key, value in pairs:
        if key in result or key in FORBIDDEN:
            raise ValueError
        result[key] = value
    return result


def _constant(_value: str) -> None:
    raise ValueError


def parse_json(data: bytes) -> object:
    """UTF-8 estrito, sem BOM, duplicatas, NaN ou Infinity."""
    result: object = None
    valid = False
    try:
        text = data.decode("utf-8")
        result = json.loads(
            text, object_pairs_hook=_pairs, parse_constant=_constant, parse_float=_constant
        )
        valid = True
    except (ValueError, RecursionError):
        pass
    if not valid:
        raise InvalidPresentationError("Invalid presentation.")
    return result


def _need(condition: bool) -> None:
    if not condition:
        raise InvalidPresentationError("Invalid presentation.")


def _segments(path: list[Segment]) -> None:
    _need(len(path) <= MAX_DEPTH)
    for part in path:
        _need(
            (isinstance(part, str) and bool(part) and part not in FORBIDDEN)
            or (type(part) is int and 0 <= part <= SAFE_INTEGER)
        )


def _refs(shape: Shape) -> list[str]:
    if isinstance(shape, Object):
        names = [field.name for field in shape.fields]
        _need(len(names) == len(set(names)))
        _need(all(name and name not in FORBIDDEN for name in names))
        return [field.ref for field in shape.fields]
    if isinstance(shape, Sequence):
        return [shape.item]
    if isinstance(shape, Union):
        _need(shape.tag not in FORBIDDEN and bool(shape.tag))
        values = [json.dumps(item.value) for item in shape.variants]
        _need(len(values) == len(set(values)))
        return [item.ref for item in shape.variants]
    if isinstance(shape, Text | Integer):
        _need(shape.min <= shape.max)
    if isinstance(shape, Choice):
        values = [json.dumps(item) for item in shape.choices]
        _need(len(values) == len(set(values)))
    return []


def _graph(models: dict[str, Definition]) -> None:
    edges = {key: _refs(value.shape) for key, value in models.items()}
    heights: dict[str, int] = {}

    def visit(key: str, ancestors: frozenset[str]) -> int:
        _need(key in models and key not in ancestors and len(ancestors) < MAX_DEPTH)
        if key in heights:
            return heights[key]
        height = 1 + max((visit(ref, ancestors | {key}) for ref in edges[key]), default=0)
        _need(height <= MAX_DEPTH)
        heights[key] = height
        return height

    for key in models:
        visit(key, frozenset())
    for definition in models.values():
        shape = definition.shape
        if isinstance(shape, Object):
            for field in shape.fields:
                _default(field.default, models[field.ref].shape, models)
        if isinstance(shape, Union):
            for variant in shape.variants:
                target = models[variant.ref].shape
                _need(isinstance(target, Object))
                if isinstance(target, Object):
                    tags = [f for f in target.fields if f.name == shape.tag and f.required]
                    _need(len(tags) == 1)
                    node = models[tags[0].ref].shape
                    _need(isinstance(node, Choice) and node.choices == [variant.value])


def _default(value: Scalar, shape: Shape, models: dict[str, Definition]) -> None:
    if value is None:
        return
    while isinstance(shape, Sequence) and shape.type == "nullable":
        shape = models[shape.item].shape
    if isinstance(shape, Boolean):
        _need(type(value) is bool)
    elif isinstance(shape, Integer):
        _need(type(value) is int and shape.min <= value <= shape.max)
    elif isinstance(shape, Text):
        _need(isinstance(value, str))
        if isinstance(value, str):
            _need(shape.min <= len(value) <= shape.max and value.startswith(shape.prefix))
            _need(
                not shape.alphabet or all(c in shape.alphabet for c in value[len(shape.prefix) :])
            )
    elif isinstance(shape, Choice):
        _need(any(type(v) is type(value) and v == value for v in shape.choices))
    else:
        _need(False)


def _path(model: str, path: list[Segment], models: dict[str, Definition]) -> Shape:
    _segments(path)
    _need(model in models)
    for part in path:
        shape = models[model].shape
        while isinstance(shape, Sequence) and shape.type == "nullable":
            shape = models[shape.item].shape
        if isinstance(shape, Object) and isinstance(part, str):
            fields = [f for f in shape.fields if f.name == part]
            _need(len(fields) == 1)
            model = fields[0].ref
        elif isinstance(shape, Sequence) and shape.type == "list" and type(part) is int:
            model = shape.item
        else:
            _need(False)
    return models[model].shape


def _destination(call: Call) -> None:
    _need(call.path.startswith("/admin/v1/"))
    _need(not any(char in call.path for char in ("?", "#", "%", "\\", "//")))
    segments = call.path.split("/")[3:]
    _need(all(part and part not in {".", ".."} for part in segments))
    slots = [part for part in segments if "{" in part or "}" in part]
    _need(len(slots) <= 1)
    if slots:
        _need(call.identity not in FORBIDDEN)
        _need(call.identity is not None and slots == ["{" + str(call.identity) + "}"])
        _need(bool(re.fullmatch(r"[a-z_]+", str(call.identity))))
    else:
        _need(call.identity is None)


def _tree(value: object, depth: int) -> None:
    _need(depth <= MAX_DEPTH)
    if type(value) is int:
        _need(-SAFE_INTEGER <= value <= SAFE_INTEGER)
    if isinstance(value, dict):
        _need(not FORBIDDEN.intersection(value))
        for child in value.values():
            _tree(child, depth + 1)
    elif isinstance(value, list):
        for child in value:
            _tree(child, depth + 1)


def validate_presentation(data: bytes) -> Presentation:
    """Valida formato e referencias, sem importar schemas/HTTP ou ler runtime."""
    _need(len(data) <= MAX_PRESENTATION_BYTES)
    raw = parse_json(data)
    _tree(raw, 0)
    _need(isinstance(raw, dict) and type(raw.get("format")) is int)
    result: Presentation | None = None
    with suppress(ValidationError):
        result = Presentation.model_validate(raw)
    if result is None:
        raise InvalidPresentationError("Invalid presentation.")
    _need(type(result.format) is int)
    identities: list[str] = []
    collections = (
        result.models,
        result.calls,
        result.views,
        result.editors,
        result.bindings,
        result.messages,
    )
    for collection in collections:
        identities.extend(item.id for item in collection)
    controls = [c for view in result.views for c in view.controls]
    controls.extend(c for editor in result.editors for c in editor.controls)
    identities.extend(control.id for control in controls)
    _need(len(controls) <= MAX_CONTROLS)
    _need(len(identities) == len(set(identities)))
    _need(all(re.fullmatch(r"[a-z][0-9]+", key) for key in identities))
    models = {item.id: item for item in result.models}
    _graph(models)
    calls = {item.id: item for item in result.calls}
    for call in result.calls:
        _destination(call)
        _need(
            call.output in models
            and call.error in models
            and (call.input is None or call.input in models)
        )
        _need((call.method == "GET") == (call.input is None))
    for view in result.views:
        _need(view.call in calls and calls[view.call].method == "GET")
        _need(all(action in calls for action in view.actions))
    for control in controls:
        leaf = _path(control.model, control.path, models)
        _default(control.default, leaf, models)
        if control.condition is not None:
            _path(control.model, control.condition.path, models)
        for projection in control.projections:
            for path in projection.paths:
                _path(control.model, path, models)
            _segments(projection.target)
    for editor in result.editors:
        _need(editor.model in models)
    for binding in result.bindings:
        _path(binding.model, binding.path, models)
    return result

File: frontend/src/protocol.js
/** @param {unknown} value @returns {value is Record<string, unknown>} */
function record(value) {
  return typeof value === "object" && value !== null && !Array.isArray(value)
    && (Object.getPrototypeOf(value) === Object.prototype || Object.getPrototypeOf(value) === null);
}
/** @param {unknown} value @returns {value is unknown[]} */
function sequence(value) { return Array.isArray(value); }
/** @param {unknown} value @returns {value is string[]} */
function strings(value) { return sequence(value) && value.every(v => typeof v === "string"); }
const denied = new Set(["__proto__", "constructor", "prototype"]);
/** @param {unknown} value @param {number} depth @param {Set<object>} seen @returns {boolean} */
function safe(value, depth, seen) {
  if (depth > 16) return false;
  if (typeof value === "number") return Number.isSafeInteger(value);
  if (value === null || typeof value === "string" || typeof value === "boolean") return true;
  if (!record(value) && !sequence(value)) return false;
  if (seen.has(value)) return false;
  seen.add(value);
  const ok = Object.keys(value).every(k => !denied.has(k))
    && Object.values(value).every(v => safe(v, depth + 1, seen));
  seen.delete(value);
  return ok;
}
/** @param {unknown} value @param {unknown} node @param {Record<string,unknown>} defs @param {number} depth @returns {boolean} */
function accepts(value, node, defs, depth) {
  if (!record(node) || depth > 32) return false;
  if (typeof node.$ref === "string") return accepts(value, defs[node.$ref.split("/").at(-1) ?? ""], defs, depth+1);
  if ("const" in node && value !== node.const) return false;
  if (sequence(node.enum) && !node.enum.some(v => v === value)) return false;
  if (sequence(node.anyOf) && !node.anyOf.some(n => accepts(value,n,defs,depth+1))) return false;
  if (sequence(node.oneOf) && node.oneOf.filter(n => accepts(value,n,defs,depth+1)).length !== 1) return false;
  if (node.type === "null" && value !== null) return false;
  if (node.type === "boolean" && typeof value !== "boolean") return false;
  if (node.type === "integer" && (typeof value !== "number" || !Number.isSafeInteger(value))) return false;
  if (node.type === "string" && typeof value !== "string") return false;
  if (typeof value === "number") {
    if (typeof node.minimum === "number" && value < node.minimum) return false;
    if (typeof node.maximum === "number" && value > node.maximum) return false;
  }
  if (typeof value === "string") {
    if (typeof node.minLength === "number" && [...value].length < node.minLength) return false;
    if (typeof node.maxLength === "number" && [...value].length > node.maxLength) return false;
  }
  if (node.type === "array") {
    if (!sequence(value)) return false;
    if (typeof node.minItems === "number" && value.length < node.minItems) return false;
    if (typeof node.maxItems === "number" && value.length > node.maxItems) return false;
    if (!value.every(v => accepts(v,node.items,defs,depth+1))) return false;
  }
  if (node.type === "object") {
    if (!record(value) || !record(node.properties)) return false;
    const props = node.properties;
    if (strings(node.required) && !node.required.every(k => Object.hasOwn(value,k))) return false;
    if (!Object.keys(value).every(k => Object.hasOwn(props,k))) return false;
    if (!Object.entries(value).every(([k,v]) => accepts(v,props[k],defs,depth+1))) return false;
  }
  return true;
}
/** @param {unknown} value @returns {value is Record<string,unknown>[]} */
function records(value) { return sequence(value) && value.every(record); }
/** @param {unknown} value @param {unknown} descriptor @returns {boolean} */
function inspect(value, descriptor) {
  if (!record(descriptor) || !record(descriptor.$defs) || !safe(value,0,new Set())) return false;
  if (!accepts(value,descriptor,descriptor.$defs,0) || !record(value)) return false;
  if (!records(value.models) || !records(value.calls) || !records(value.views)
    || !records(value.editors) || !records(value.bindings) || !records(value.messages)) return false;
  const all = [...value.models,...value.calls,...value.views,...value.editors,...value.bindings,...value.messages];
  /** @type {Record<string,unknown>[]} */ const controls=[];
  for (const owner of [...value.views,...value.editors]) {
    if (!records(owner.controls)) return false;
    controls.push(...owner.controls);
  }
  if (controls.length > 512) return false;
  all.push(...controls);
  const ids=all.map(v=>v.id);
  if (!ids.every(v => typeof v === "string" && /^[a-z][0-9]+$/.test(v)) || new Set(ids).size !== ids.length) return false;
  /** @type {Map<string,Record<string,unknown>>} */ const models=new Map();
  for (const item of value.models) {
    if (typeof item.id !== "string" || !record(item.shape)) return false;
    models.set(item.id,item.shape);
  }
  /** @type {Map<string,number>} */ const heights=new Map();
  /** @param {string} key @param {Set<string>} seen @returns {number} */
  function visit(key,seen) {
    const node=models.get(key);
    if (!node || seen.has(key) || seen.size >= 16) return 0;
    const cached=heights.get(key);
    if(cached !== undefined) return cached;
    const next=new Set([...seen,key]);
    /** @type {unknown[]} */ let refs=[];
    if (node.type === "object") {
      if (!records(node.fields)) return 0;
      const names=node.fields.map(f=>f.name);
      if (!names.every(n=>typeof n === "string" && n.length > 0 && !denied.has(n)) || new Set(names).size !== names.length) return 0;
      refs=node.fields.map(f=>f.ref);
    }
    if (node.type === "list" || node.type === "nullable") refs=[node.item];
    if (node.type === "union") {
      if (typeof node.tag !== "string" || !node.tag || denied.has(node.tag) || !records(node.variants)) return 0;
      const values=node.variants.map(v=>JSON.stringify(v.value));
      if (new Set(values).size !== values.length) return 0;
      refs=node.variants.map(v=>v.ref);
      for (const variant of node.variants) {
        if (typeof variant.ref !== "string") return 0;
        const target=models.get(variant.ref);
        if (!target || target.type !== "object" || !records(target.fields)) return 0;
        const link=target.fields.find(f=>f.name === node.tag && f.required === true);
        if (!link || typeof link.ref !== "string") return 0;
        const tag=models.get(link.ref);
        if (!tag || tag.type !== "enum" || !sequence(tag.choices) || tag.choices.length !== 1 || tag.choices[0] !== variant.value) return 0;
      }
    }
    if (typeof node.min === "number" && typeof node.max === "number" && node.min > node.max) return 0;
    if (sequence(node.choices)) {
      const values=node.choices.map(v=>JSON.stringify(v));
      if (new Set(values).size !== values.length) return 0;
    }
    const levels=refs.map(ref=>typeof ref === "string" ? visit(ref,next) : 0);
    if(levels.some(v=>v===0)) return 0;
    const height=1+Math.max(0,...levels);
    if(height>16) return 0;
    heights.set(key,height);
    return height;
  }
  if (![...models.keys()].every(key=>visit(key,new Set()))) return false;
  /** @param {unknown} key @param {unknown} parts @returns {Record<string,unknown> | undefined} */
  function leaf(key,parts) {
    if (typeof key !== "string" || !sequence(parts) || parts.length > 16 || !models.has(key)) return undefined;
    let node=models.get(key);
    for (const part of parts) {
      while (node?.type === "nullable" && typeof node.item === "string") node=models.get(node.item);
      if (typeof part === "string" && part && !denied.has(part) && node?.type === "object" && records(node.fields)) {
        const field=node.fields.find(f=>f.name === part);
        if (!field || typeof field.ref !== "string") return undefined;
        node=models.get(field.ref);
      } else if (typeof part === "number" && Number.isSafeInteger(part) && part >= 0 && node?.type === "list" && typeof node.item === "string") node=models.get(node.item);
      else return undefined;
    }
    return node;
  }
  /** @param {unknown} key @param {unknown} parts */
  function linked(key,parts) { return leaf(key,parts) !== undefined; }
  /** @param {unknown} value @param {Record<string,unknown> | undefined} node @returns {boolean} */
  function fits(value,node) {
    if(value === null || value === undefined) return true;
    while(node?.type === "nullable" && typeof node.item === "string") node=models.get(node.item);
    if(!node) return false;
    if(node.type === "boolean") return typeof value === "boolean";
    if(node.type === "integer") return typeof value === "number" && Number.isSafeInteger(value) && value >= (typeof node.min === "number" ? node.min : 0) && value <= (typeof node.max === "number" ? node.max : Number.MAX_SAFE_INTEGER);
    if(node.type === "enum") return sequence(node.choices) && node.choices.some(v=>v === value);
    if(node.type === "string") {
      if(typeof value !== "string") return false;
      const prefix=typeof node.prefix === "string" ? node.prefix : "";
      const alphabet=typeof node.alphabet === "string" ? node.alphabet : "";
      return [...value].length >= (typeof node.min === "number" ? node.min : 0) && [...value].length <= (typeof node.max === "number" ? node.max : Number.MAX_SAFE_INTEGER) && value.startsWith(prefix) && (!alphabet || [...value.slice(prefix.length)].every(c=>alphabet.includes(c)));
    }
    return false;
  }
  for(const node of models.values()) if(node.type === "object" && records(node.fields)) {
    for(const field of node.fields) if(typeof field.ref !== "string" || !fits(field.default,models.get(field.ref))) return false;
  }
  for(const control of controls) if(!fits(control.default,leaf(control.model,control.path))) return false;
  const calls=new Map(value.calls.map(c=>[c.id,c]));
  for (const call of value.calls) {
    if (typeof call.path !== "string" || !call.path.startsWith("/admin/v1/") || /[?#%\\]|\/\//.test(call.path)) return false;
    const parts=call.path.split("/").slice(3);
    if (parts.some(p=>!p || p === "." || p === "..")) return false;
    const slots=parts.filter(p=>p.includes("{") || p.includes("}"));
    if (slots.length > 1 || (slots.length === 0 && call.identity !== null)) return false;
    if (slots.length === 1 && (typeof call.identity !== "string" || denied.has(call.identity) || !/^[a-z_]+$/.test(call.identity) || slots[0] !== "{"+call.identity+"}")) return false;
    if (typeof call.error !== "string" || !models.has(call.error) || typeof call.output !== "string" || !models.has(call.output)) return false;
    if ((call.method === "GET") !== (call.input === null)) return false;
    if (call.input !== null && (typeof call.input !== "string" || !models.has(call.input))) return false;
  }
  for (const view of value.views) {
    if (calls.get(view.call)?.method !== "GET" || !sequence(view.actions) || !view.actions.every(a=>calls.has(a))) return false;
  }
  for (const editor of value.editors) if (typeof editor.model !== "string" || !models.has(editor.model)) return false;
  for (const control of controls) {
    if (control.projections !== undefined) {
      if(!records(control.projections)) return false;
      for (const projection of control.projections) {
        if(!sequence(projection.paths) || !projection.paths.every(p=>linked(control.model,p))) return false;
        if(!sequence(projection.target) || projection.target.length>16 || !projection.target.every(p=>(typeof p === "string" && p.length>0 && !denied.has(p)) || (typeof p === "number" && Number.isSafeInteger(p) && p>=0))) return false;
      }
    }
  }
  for (const item of [...controls,...value.bindings]) {
    if (!linked(item.model,item.path)) return false;
    if (item.condition !== null && item.condition !== undefined && (!record(item.condition) || !linked(item.model,item.condition.path))) return false;
  }
  return true;
}
export {};
